### DISCUSSION TITLE
Private Route Table Ids not found.

### DISCUSSION THREAD

**Discussion by fade2black** [aws/copilot-cli#4668] [Q&A]
Both 

`Fn::ImportValue: !Sub ${App}-${Env}-PrivateRouteTableIDs`

and 

`Fn::ImportValue: !Sub ${AWS::StackName}-PrivateRouteTableIDs` 

result in the following error:  `No export named myapp-staging-PrivateRouteTableIDs found` when I refer from an env addon.

Actually I got the same error when I tried to use other resources too (e.g., VpcId). I suspect this error occurs because the addon is a nested (child) stack which is created in parallel together with the parent stack. But since the parent stack is not ready yet the resources referred from the child stack are not found. Usually we deal with it using `DpendsOn`. Does anyone know how to deal with it? In case of VPC id I resolved it by specifying 
```yaml
Parameters:
  VpcId: !Ref VPC
  ...
```
inside `addons.parameters.yml`. But when I specify `PrivateRouteTable1` it says "PrivateRouteTable1 not found"  Do I miss something?

**Comment by bvtujo** [comment]
It seems like the `!Ref VPC` call works because the VPC already exists, and the `!Ref PrivateRouteTable1` call doesn't work becasue the route table doesn't exist. Is this the first service in your environment, or the first with a `private` placement? Or is this the environment addons stack with no workloads deployed?

By default, when you first deploy an environment we don't create NAT gateways or any of the routing infrastructure for web traffic to reach the private subnet. Mostly this is for cost reasons; NAT gateways are expensive for small projects and we want to minimize costs for customers until they evolve to a place in their infrastructure where they can opt in. Instead, we only create that infrastructure when you specify `private` task placement in one or more of your services. 

That's done by attaching a `CreateNATGateways` condition to the private routing infrastructure/outputs and setting that condition based on a `NATWorkloads` input parameter; you can verify this in the Environment CFN template. When you deploy a service with private placement, the service's Environment Controller resource causes a redeployment of the Environment template with the `NATWorkloads` parameter updated, causing the routing infrastructure to be created.

All of this is to say that it's not working because the PrivateRouteTable1 resource doesn't exist yet. 

You can work around it by doing one of the following:
1. If this is a service or job addons stack: deploy a placeholder workload with `placement: private` into the environment before deploying this workload.
2. If this is an environment stack: deploy the environment first without addons, then deploy a placeholder workload as above, then define the addons stack and redeploy the environment.
